pallets/pools/src/tranche.rs

```diff
@@ -0,0 +1,386 @@
+					min_risk_buffer: input.min_risk_buffer.unwrap_or(Perquintill::zero()),
+					// seniority increases as index since the order is from junior to senior
+					seniority: input
+						.seniority
+						.unwrap_or(id.try_into().expect("MaxTranches is u32")),
+					outstanding_invest_orders: Zero::zero(),
+					outstanding_redeem_orders: Zero::zero(),
+
+					debt: Zero::zero(),
+					reserve: Zero::zero(),
+					ratio: Perquintill::zero(),
+					last_updated_interest: now,
+					_phantom: Default::default(),
+				})
+				.collect(),
+		)
+	}
+
+	pub fn num_tranches(&self) -> usize {
+		self.0.len()
+	}
+
+	pub fn into_tranches(self) -> Vec<Tranche<Balance, Rate, Weight>> {
+		self.0
+	}
+
+	pub fn as_tranche_slice(&self) -> &[Tranche<Balance, Rate, Weight>] {
+		self.0.as_slice()
+	}
+
+	pub fn as_mut_tranche_slice(&mut self) -> &mut [Tranche<Balance, Rate, Weight>] {
+		self.0.as_mut_slice()
+	}
+
+	pub fn supplies(&self) -> Result<Vec<Balance>, DispatchError> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.debt.checked_add(&tranche.reserve))
+			.collect::<Option<Vec<_>>>()
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn acc_supply(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| {
+					acc.checked_add(&tranche.debt)
+						.and_then(|acc| acc.checked_add(&tranche.reserve))
+				})
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn investments(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_invest_orders)
+			.collect()
+	}
+
+	pub fn acc_investments(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| acc.checked_add(&tranche.outstanding_invest_orders))
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn redemptions(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_redeem_orders)
+			.collect()
+	}
+
+	pub fn acc_redemptions(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| acc.checked_add(&tranche.outstanding_redeem_orders))
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn calculate_weights(&self) -> Vec<(Weight, Weight)> {
+		let n_tranches: u32 = self.0.len().try_into().expect("MaxTranches is u32");
+		let redeem_starts = 10u128.checked_pow(n_tranches).unwrap_or(u128::MAX);
+
+		self.0
+			.iter()
+			.map(|tranche| {
+				(
+					10u128
+						.checked_pow(
+							n_tranches
+								.checked_sub(tranche.seniority)
+								.unwrap_or(u32::MAX),
+						)
+						.unwrap_or(u128::MAX)
+						.into(),
+					redeem_starts
+						.checked_mul(10u128.pow(tranche.seniority.saturating_add(1)).into())
+						.unwrap_or(u128::MAX)
+						.into(),
+				)
+			})
+			.collect()
+	}
+
+	pub fn risk_buffers(&self) -> Vec<Perquintill> {
```

Comment:
This should be called `min_risk_buffers`

Reply:
Done.